crates/ruff_server/src/session/workspace.rs

```diff
@@ -0,0 +1,254 @@
+            },
+            |workspace| {
+                settings::ResolvedClientSettings::with_workspace(
+                    &workspace.settings,
+                    global_settings,
+                )
+            },
+        )
+    }
+
+    fn workspace_for_url(&self, url: &Url) -> Option<&Workspace> {
+        Some(self.entry_for_url(url)?.1)
+    }
+
+    fn workspace_for_url_mut(&mut self, url: &Url) -> Option<&mut Workspace> {
+        Some(self.entry_for_url_mut(url)?.1)
+    }
+
+    fn entry_for_url(&self, url: &Url) -> Option<(&Path, &Workspace)> {
+        let path = url.to_file_path().ok()?;
+        self.0
+            .range(..path)
+            .next_back()
+            .map(|(path, workspace)| (path.as_path(), workspace))
+    }
+
+    fn entry_for_url_mut(&mut self, url: &Url) -> Option<(&Path, &mut Workspace)> {
+        let path = url.to_file_path().ok()?;
+        self.0
+            .range_mut(..path)
+            .next_back()
+            .map(|(path, workspace)| (path.as_path(), workspace))
+    }
+}
+
+impl Workspace {
+    pub(crate) fn new(root: &Url, settings: ClientSettings) -> crate::Result<(PathBuf, Self)> {
+        let path = root
+            .to_file_path()
+            .map_err(|()| anyhow!("workspace URL was not a file path!"))?;
+
+        let workspace = Self {
+            open_documents: OpenDocuments::new(&path),
+            settings,
+        };
+
+        Ok((path, workspace))
+    }
+
+    fn reload_ruff_settings(&mut self, root: &Path) {
+        self.open_documents.reload_ruff_settings(root);
+    }
+}
+
+impl OpenDocuments {
+    fn new(path: &Path) -> Self {
+        Self {
+            documents: FxHashMap::default(),
+            ruff_settings_index: RuffSettingsIndex::new(path),
+        }
+    }
+
+    fn snapshot(&self, url: &Url) -> Option<DocumentRef> {
+        let path = url
+            .to_file_path()
+            .expect("document URL should convert to file path: {url}");
+        let document_settings = self.ruff_settings_index.get(&path).clone();
+        Some(self.documents.get(url)?.make_ref(document_settings))
+    }
+
+    fn controller(&mut self, url: &Url) -> Option<&mut DocumentController> {
+        self.documents.get_mut(url)
+    }
+
+    fn open(&mut self, url: &Url, contents: String, version: DocumentVersion) {
+        if self
+            .documents
+            .insert(url.clone(), DocumentController::new(contents, version))
+            .is_some()
+        {
+            tracing::warn!("Opening document `{url}` that is already open!");
+        }
+    }
+
+    fn close(&mut self, url: &Url) -> crate::Result<()> {
+        let Some(_) = self.documents.remove(url) else {
+            return Err(anyhow!(
+                "Tried to close document `{url}`, which was not open"
+            ));
+        };
+        Ok(())
+    }
+
+    fn reload_ruff_settings(&mut self, root: &Path) {
+        self.ruff_settings_index.reload(root);
+    }
```

Comment:
nit: we could potentially drop `reload` and `clear` from the index and re-create the settings with the same method:

```diff
diff --git a/crates/ruff_server/src/session/workspace.rs b/crates/ruff_server/src/session/workspace.rs
index 10d2b8a56..068236f8c 100644
--- a/crates/ruff_server/src/session/workspace.rs
+++ b/crates/ruff_server/src/session/workspace.rs
@@ -210,7 +210,7 @@ impl OpenDocuments {
     }
 
     fn reload_ruff_settings(&mut self, root: &Path) {
-        self.ruff_settings_index.reload(root);
+        self.ruff_settings_index = RuffSettingsIndex::new(root);
     }
 }
 
diff --git a/crates/ruff_server/src/session/workspace/ruff_settings.rs b/crates/ruff_server/src/session/workspace/ruff_settings.rs
index 6bff2a975..376c850a5 100644
--- a/crates/ruff_server/src/session/workspace/ruff_settings.rs
+++ b/crates/ruff_server/src/session/workspace/ruff_settings.rs
@@ -38,20 +38,9 @@ impl std::fmt::Display for RuffSettings {
 
 impl RuffSettingsIndex {
     pub(super) fn new(path: &Path) -> Self {
-        let mut index = Self {
-            index: BTreeMap::default(),
-            fallback: Arc::default(),
-        };
+        let mut index = BTreeMap::default();
 
-        index.reload(path);
-
-        index
-    }
-
-    pub(super) fn reload(&mut self, root: &Path) {
-        self.clear();
-
-        for directory in WalkDir::new(root)
+        for directory in WalkDir::new(path)
             .into_iter()
             .filter_map(Result::ok)
             .filter(|entry| entry.file_type().is_dir())
@@ -65,7 +54,7 @@ impl RuffSettingsIndex {
                 ) else {
                     continue;
                 };
-                self.index.insert(
+                index.insert(
                     directory,
                     Arc::new(RuffSettings {
                         linter: settings.linter,
@@ -74,6 +63,11 @@ impl RuffSettingsIndex {
                 );
             }
         }
+
+        Self {
+            index,
+            fallback: Arc::default(),
+        }
     }
 
     pub(super) fn get(&self, document_path: &Path) -> &Arc<RuffSettings> {
@@ -90,10 +84,6 @@ impl RuffSettingsIndex {
 
         &self.fallback
     }
-
-    fn clear(&mut self) {
-        self.index.clear();
-    }
 }
 
 struct LSPConfigTransformer;
```